pipeline.go

```diff
@@ -0,0 +1,187 @@
+package rules
+
+import (
+	"errors"
+)
+
+// StageRegistry is a mapping of stage names to stage functions
+type StageRegistry map[string]StageFunc
+
+const (
+	StageSpawnFoodStandard    = "spawn_food.standard"
+	StageGameOverStandard     = "game_over.standard"
+	StageStarvationStandard   = "starvation.standard"
+	StageFeedSnakesStandard   = "feed_snakes.standard"
+	StageMovementStandard     = "movement.standard"
+	StageHazardDamageStandard = "hazard_damage.standard"
+	StageEliminationStandard  = "elimination.standard"
+
+	StageGameOverSoloSnake                   = "game_over.solo_snake"
+	StageGameOverBySquad                     = "game_over.by_squad"
+	StageSpawnFoodNoFood                     = "spawn_food.no_food"
+	StageSpawnHazardsShrinkMap               = "spawn_hazards.shrink_map"
+	StageEliminationResurrectSquadCollisions = "elimination.resurrect_squad_collisions"
+	StageModifySnakesAlwaysGrow              = "modify_snakes.always_grow"
+	StageMovementWrapBoundaries              = "movement.wrap_boundaries"
+	StageModifySnakesShareAttributes         = "modify_snakes.share_attributes"
+)
+
+// globalRegistry is a global, default mapping of stage names to stage functions.
+// It can be extended by plugins through the use of registration functions.
+// Plugins that wish to extend the available game stages should call RegisterPipelineStageError
+// to add additional stages.
+var globalRegistry = StageRegistry{
+	StageSpawnFoodNoFood:                     RemoveFoodConstrictor,
+	StageSpawnFoodStandard:                   SpawnFoodStandard,
+	StageGameOverSoloSnake:                   GameOverSolo,
+	StageGameOverBySquad:                     GameOverSquad,
+	StageGameOverStandard:                    GameOverStandard,
+	StageHazardDamageStandard:                DamageHazardsStandard,
+	StageSpawnHazardsShrinkMap:               PopulateHazardsRoyale,
+	StageStarvationStandard:                  ReduceSnakeHealthStandard,
+	StageEliminationResurrectSquadCollisions: ResurrectSnakesSquad,
+	StageFeedSnakesStandard:                  FeedSnakesStandard,
+	StageEliminationStandard:                 EliminateSnakesStandard,
+	StageModifySnakesAlwaysGrow:              GrowSnakesConstrictor,
+	StageMovementStandard:                    MoveSnakesStandard,
+	StageMovementWrapBoundaries:              MoveSnakesWrapped,
+	StageModifySnakesShareAttributes:         ShareAttributesSquad,
+}
+
+// RegisterPipelineStage adds a stage to the registry.
+// If a stage has already been mapped it will be overwritten by the newly
+// registered function.
+func (sr StageRegistry) RegisterPipelineStage(s string, fn StageFunc) {
+	sr[s] = fn
+}
+
+// RegisterPipelineStageError adds a stage to the registry.
+// If a stage has already been mapped an error will be returned.
```

Comment:
Good idea to check for duplicates, but what would calling code do with this error? I'm guessing this will normally be called in an `init` function alongside newly defined StageFuncs, where an error should just be treated as unrecoverable, right?

One pattern I've seen for things like metrics registries is something like this:
```
func MustRegisterPipelineStage(s string, fn StageFunc) {
```
where instead of returning an error, it'll just panic if there's a conflict. I don't think that's a bad use of panic, since we want to avoid code that could possibly trigger it from being merged in the first place.

Reply:
I considered panic, but I felt it was too unsafe and that the issue of duplicates was not critical. I thought that it made sense for the API user to decide whether duplicates were a problem, so I added two registration methods:

`RegisterPipelineStageError` (errors on duplicates)
and
`RegisterPipelineStage` (no error on duplicates)

I have seen the pattern you mentioned (`Must*` methods), but I don't think they are a good idea for libraries. Libraries should not panic and only `main` package code, which could decide this error is unrecoverable or they could decide to log.

I feel that returning an error is pretty much the same convenience with more flexibility and is safer.

Reply:
> Libraries should not panic and only main package code should decide if an error is unrecoverable or should just logged/whatever
Fair enough, IMO it's not quite that clear cut though. If code in rules is registering two stages under the same name, that should result in rules blowing up in a way that avoids it ever being released, and failing that, immediately obvious that it's a broken release if included into a main package.

I'm not 100% set on that though since the happy path will work regardless. If the `RegisterPipeline*` functions will be called internally within rules, how do you see the error being handled instead?

Reply:
Actually, since I seem to be such a big fan of looking at the `database/sql` package for inspiration, it does suggest we should panic in this type of situation. See the `Register` function in https://cs.opensource.google/go/go/+/master:src/database/sql/sql.go .

Assuming we want to make this change, would you want me to also remove the `RegisterPipelineStage` method? If I understand your thinking, this would ensure that all contributors would be checking for duplicates.